### PR TITLE
Added static code tests using MyPy

### PR DESCRIPTION
**Pull request by BvB93**
See https://github.com/SCM-NV/qmflows/issues/176.

[Mypy](https://github.com/python/mypy) is an optional static type checker for Python that aims to combine the benefits of dynamic (or "duck") typing and static typing. [Mypy](https://github.com/python/mypy) combines the expressive power and convenience of Python with a powerful type system and compile-time type checking. [Mypy](https://github.com/python/mypy) type checks standard Python programs; run them using any Python VM with basically no runtime overhead. 

Major
------
* Introduced tests using MyPy.
* Refined existing type hints.

Minor
------
* Fixed the occasional ``AttributeError`` that might be raised due to unexpected ``None`` values.

Open question
---------------
* ~~Should these newly introduced test raise an ``Exception`` or issue a ``Warning``?~~ 
  Use ``Warnings`` for the moment.

### CODE REVIEW

**Comment by felipeZ** [comment]
I am in favor of typing everything but the library is still far from having the most function/methods properly typed. So, I suggest raising warning for some time until we introduce more types

**Comment by BvB93** [comment]
> I am in favor of typing everything but the library is still far from having the most function/methods properly typed. So, I suggest raising warning for some time until we introduce more types

Agreed; at the very least we'll have to be _very_ picky about when to raise an ``Exception``.

**Comment by felipeZ** [comment]
Look at the [codecov](https://docs.codecov.io/docs/graphs#sunburst) report we are having other colors than red :heart_eyes:

**Comment by BvB93** [comment]
> Look at the [codecov](https://docs.codecov.io/docs/graphs#sunburst) report we are having other colors than red 😍

**gasp**!
It's beautiful!

File: setup.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ def readme():
                       'pyparsing', 'pyyaml>=5.1', 'filelock'],
     extras_require={
         'test': ['pytest', 'pytest-cov', 'pytest-mock', 'typing_extensions',
-                 'assertionlib', 'pycodestyle'],
+                 'assertionlib', 'pycodestyle', 'mypy'],
```
Comment (felipeZ) [inline setup.py]: A seemly small import, a gigantic improvement in mental sanity!

File: src/qmflows/components/operations.py [hunk 2]
```diff
@@ -20,7 +20,7 @@ def find_first_job(pred: Predicate,
                    packagelist: Iterable[Package],
                    settings: Settings,
                    molecule: Molecule,
-                   job_name: str, **kwargs: Any) -> Result:
+                   job_name: str, **kwargs: Any) -> Optional[Result]:
```
Comment (felipeZ) [inline src/qmflows/components/operations.py]: It is indeed easier to reason about the code

File: src/qmflows/cp2k_utils.py [hunk 3]
```diff
@@ -362,9 +364,6 @@ def _(unit: Optional[str]) -> List[str]:
         return [f'[{unit}] {{}}']
 
 
-NDFrame: type = pd.DataFrame.__bases__[0]
```
Comment (felipeZ) [inline src/qmflows/cp2k_utils.py]: You make me happy every time that you delete a magic method/property :+1:

File: src/qmflows/parsers/cp2KParser.py [hunk 4]
```diff
@@ -57,9 +58,12 @@ def parse_cp2k_warnings(file_name: PathLike,
     return warnings or None
 
 
+#: A generator yielding 2-tuples with warning messages and warning types.
+WarnGenerator = Generator[Tuple[str, Type[Warning]], None, None]
```
Comment (felipeZ) [inline src/qmflows/parsers/cp2KParser.py]: I don't understand why 2 `None`s.  Would be a single `None` good enough?

File: src/qmflows/parsers/cp2KParser.py [hunk 5]
```diff
@@ -69,18 +73,22 @@ def assign_warning(warning_type: Type[Warning], msg: str,
...
     :returns: NamedTuple containing the Eigenvalues and the Coefficients
     """
-    file_out = fnmatch.filter(os.listdir(plams_dir), '*out')[0]
+    plams_dir = Path(plams_dir) if plams_dir else Path(os.getcwd())
+
     file_in = fnmatch.filter(os.listdir(plams_dir), '*in')[0]
-    path_in, path_out = [os.path.join(plams_dir, x)
-                         for x in [file_in, file_out]]
+    file_out = fnmatch.filter(os.listdir(plams_dir), '*out')[0]
+
```
Comment (felipeZ) [inline src/qmflows/parsers/cp2KParser.py]: the '/' operator makes things much cleaner

File: src/qmflows/parsers/cp2KParser.py [hunk 6]
```diff
@@ -279,7 +290,7 @@ def readCp2KBasis(path: PathLike) -> Tuple2List:
     rss = [swapCoeff(*args) for args in zip(nCoeffs, coefficients)]
     tss = [headTail(xs) for xs in rss]
     basisData = [AtomBasisData(xs[0], xs[1]) for xs in tss]
-    basiskey = [AtomBasisKey(*rs) for rs in zip(atoms, names, formats)]
```
Comment (felipeZ) [inline src/qmflows/parsers/cp2KParser.py]: yeah, you are right ;)

File: src/qmflows/parsers/generic_parsers.py [hunk 7]
```diff
@@ -34,22 +44,23 @@ def awk_file(filename, script='', progfile=None):
...
         except ValueError:
             try:
                 v = float(i)
             except ValueError:
                 v = i
         result.append(v)
-    if len(result) == 1:
-        result = result[0]
-    return result
+    return result[0] if len(result) == 1 else result
```
Comment (felipeZ) [inline src/qmflows/parsers/generic_parsers.py]: top!

File: test/test_mypy.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+_ROOT = Path(__file__).parts[:-2]
+PACKAGE = join(*_ROOT, 'src', 'qmflows')
+INI = str(PATH / 'mypy.ini')
+
+ACTION = frozenset({'raise', 'warn', 'ignore'})
+Action = Literal[tuple(ACTION)]  #: Type annotation for the 'action' keyword.
+
+
+def test_mypy(action: Action = 'warn') -> None:
```
Comment (felipeZ) [inline test/test_mypy.py]: This is indeed a very fine addition to improve the code quality, I hope that we can soon enforce it!

File: src/qmflows/utils.py [hunk 9]
```diff
@@ -61,7 +61,7 @@ def wrapper(settings, key, value, mol, **kwargs):
     return wrapper
 
 
-def file_to_context(file: Union[PathLike, IO],
+def file_to_context(file: Union[int, PathLike, IO],
```
Comment (felipeZ) [inline src/qmflows/utils.py]: why an `Int`?

File: src/qmflows/cp2k_utils.py (hunk 3 above)
Comment (BvB93) [reply src/qmflows/cp2k_utils.py]: But I like magic :(

File: src/qmflows/parsers/cp2KParser.py (hunk 4 above)
Comment (BvB93) [reply src/qmflows/parsers/cp2KParser.py]: The [``Generator``](https://docs.python.org/3/library/typing.html#typing.Generator) type is quite picky about its signature.
It requires _exactly_ three values:
1. The ``yield`` type.
2. The ``send`` type.
3. The ``return`` type.

File: src/qmflows/parsers/cp2KParser.py (hunk 5 above)
Comment (BvB93) [reply src/qmflows/parsers/cp2KParser.py]: Yay, pathlib!

File: src/qmflows/utils.py (hunk 9 above)
Comment (BvB93) [reply src/qmflows/utils.py]: I've no clue, but apparently ``open()`` can accept an ``int`` as valid argument.

File: src/qmflows/parsers/cp2KParser.py (hunk 4 above)
Comment (felipeZ) [reply src/qmflows/parsers/cp2KParser.py]: I forgot the coroutine part, thanks!

**Pull request merged**